Datastore/src/Connection/Rest.php

```diff
@@ -108,6 +111,27 @@ public function rollback(array $args)
      */
     public function runQuery(array $args)
     {
+        $this->setReadTime($args);
         return $this->send('projects', 'runQuery', $args);
     }
+
+    /**
+     * @param array $args
+     */
+    private function setReadTime(array &$args)
+    {
+        if (isset($args['readOptions']['readTime'])) {
+            $time = $args['readOptions']['readTime']->getSeconds();
+            $date = date('Y/m/d H:i:s.u\Z', $time);
+            $args['readOptions']['readTime'] = new Timestamp(
+                new \DateTime($date)
+            );
+        } elseif (isset($args['transactionOptions']['readOnly']['readTime'])) {
+            $time = $args['transactionOptions']['readOnly']['readTime']->getSeconds();
+            $date = date('Y/m/d H:i:s.u\Z', $time);
+            $args['transactionOptions']['readOnly']['readTime'] = new Timestamp(
+                new \DateTime($date)
+            );
+        }
+    }
```

Comment:
There's room to shorten this code, so that there's no repetition. Something like below (but might not work as-is):

```
function setReadTime(array &$args) {
    $options = ['readOptions', 'transactionOptions']['readOnly'];
    foreach ($options as $option) {
        if (isset($args[$option]['readTime'])) {
            $time = $args[$option]['readTime']->getSeconds();
            $date = date('Y/m/d H:i:s.u\Z', $time);
            $args[$option]['readTime'] = new Timestamp(new \DateTime($date));
        }
    }
}
```

Reply:
I don't see how the above code would work, as the `$options` line is not valid PHP. The code could pass by reference like this: 

```php
        $readTime = null;
        if (isset($args['readOptions']['readTime'])) {
            $readTime =& $args['readOptions']['readTime'];
        } elseif (isset($args['transactionOptions']['readOnly']['readTime'])) {
            $readTime =& $args['transactionOptions']['readOnly']['readTime'];
        }
        if ($readTime) {
            $time = $readTime->getSeconds();
            $date = date('Y/m/d H:i:s.u\Z', $time);
            $readTime = new Timestamp(new \DateTime($date));
        }
```
But I'm not convinced this way is better / cleaner / more readable. I typically stay away from passing variables by reference. I would prefer something like this:

```php
    public function runQuery(array $args)
    {
        $argsWithReadTime = $this->setReadTime($args);
        return $this->send('projects', 'runQuery', $argsWithReadTime);
    }

    /**
     * @param array $args
     * @return array
     */
    private function setReadTime(array $args)
    {
        if (isset($args['readOptions']['readTime'])) {
            $this->setReadTimeOption($args['readOptions']['readTime']);
        } elseif (isset($args['transactionOptions']['readOnly']['readTime'])) {
            $this->setReadTimeOption($args['transactionOptions']['readOnly']['readTime']);
        }
        return $args;
    }

    private function setReadTimeOption(&$readTime)
    {
            $date = date('Y/m/d H:i:s.u\Z', $readTime->getSeconds());
            $readTime = new Timestamp(new \DateTime($date));
    }
```

Reply:
Actually my snippet was generated by chat GPT :stuck_out_tongue_winking_eye: No wonder it wont work.
But captured the logic exactly by removing code duplication.

Reply:
Done! Thank you team for the suggestion.

Reply:
Going forward please don't generate code reviews via GPT